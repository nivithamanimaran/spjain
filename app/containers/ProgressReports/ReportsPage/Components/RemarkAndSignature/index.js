import React, { useEffect, useState } from 'react';
import { Grid, makeStyles, Box } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import DetailsContent from './Components/DetailsContent';
import RemarksContent from './Components/RemarksContent';
import SignatureContent from './Components/SignatureContent/index';
import { uploadSignature } from '../../../../../utils/ApiService';
import { KEY_SIGNATURE } from '../../../../../utils/constants';
import KenLoader from '../../../../../components/KenLoader/fullLoader';
import KenSnackBar from '../../../../../components/KenSnackbar';

const useStyles = makeStyles(theme => ({
  signContainer: {
    paddingLeft: '12px',
  },
}));

const RemarkAndSignature = props => {
  const { t } = useTranslation();
  const classes = useStyles();
  const { details, remarks, signatures, ProgramId, setStudentInfo } = props;
  const [signs, setSigns] = useState(signatures);
  const [showRemoveFileIcon, setShowRemoveFileIcon] = useState(false);

  const [showSaveFileIcon, setShowSaveFileIcon] = useState(false);
  const [droppedFiles, setDroppedFiles] = useState([]);
  const [loading, setLoading] = useState(false);
  const [openSnackbar, setOpenSnackbar] = useState(false);
  const [snackbarSeverity, setSnackbarSeverity] = useState();
  const [snackbarMessage, setSnackbarMessage] = useState();

  const handleSnackbarClose = (event, reason) => {
    if (reason === 'clickaway') return;
    setOpenSnackbar(false);
  };

  const handleSnackbarOpen = (severity, message) => {
    setOpenSnackbar(true);
    setSnackbarSeverity(severity);
    setSnackbarMessage(message);
  };

  const handleDrop = (index, acceptedFiles) => {
    setShowRemoveFileIcon(true);
    setShowSaveFileIcon(true);
    let file = acceptedFiles[0];
    if (
      KEY_SIGNATURE?.ALLOWED_FILE_TYPES?.includes(file?.type) &&
      file.size <= KEY_SIGNATURE.MAX_FILE_SIZE
    ) {
      const fileURL = URL.createObjectURL(acceptedFiles[0]);
      setDroppedFiles(acceptedFiles);
      let signaturesArray = signs;
      let obj = {
        label: signs[index]?.label,
        value: fileURL,
        newlyAdded: true,
        signType: signs[index]?.key,
      };
      signaturesArray[index] = obj;
      setSigns([...signaturesArray]);
    } else {
      //   alert(t('messages:Supported_Signature_File'));
      handleSnackbarOpen('warning', t('messages:Supported_Signature_File'));
    }
  };
  const removeFile = index => {
    let currentSigns = [...signs];
    if (currentSigns.length > 0) {
      currentSigns[index].value = '';
      currentSigns[index].newlyAdded = false;
      setSigns([...currentSigns]);
    }
  };

  const onSaveClick = index => {
    let currentSigns = [...signs];
    const signType = currentSigns[index]?.signType;
    const file = droppedFiles[0];
    if (droppedFiles && ProgramId && signType && file) {
      setLoading(true);
      uploadSignature(ProgramId, signType, file)
        .then(res => {
          setLoading(false);
          setShowRemoveFileIcon(false);
          setShowSaveFileIcon(false);
          setStudentInfo(prev => {
            return { ...prev, signatureUpload: true };
          });
        })
        .catch(err => {
          setLoading(false);
          if (err?.response?.data?.hasOwnProperty('storageErrors')) {
            alert(t('messages:Supported_Signature_File'));
          } else {
            alert(t('messages:Something_Wrong'));
          }
        });
    }
  };

  useEffect(() => {
    let signaturesCopy = [...signatures];
    const teacher = signaturesCopy.find(
      sign => sign.label === t('labels:Class_Teacher_Sign')
    );
    const parent = signaturesCopy.find(
      sign => sign.label === t('labels:Parent_Sign')
    );
    const head = signaturesCopy.find(
      sign => sign.label === t('labels:Primary_Head_Sign')
    );
    signaturesCopy = [{ ...teacher }, { ...head }, { ...parent }];
    setSigns(signaturesCopy);
  }, [signatures]);
  return (
    <>
      {loading && <KenLoader />}
      <Box data-testid="remark-and-signature-component" p={2}>
        <Grid container spacing={2}>
          <Grid item xs={12} md={8} container spacing={3}>
            {details && (
              <Grid item xs={12} md={5}>
                <DetailsContent
                  data={details}
                  title={t('headings:Remark_And_Signature_Details')}
                />
              </Grid>
            )}
            {/* <Grid item xs={12} md={7}>
              <RemarksContent
                data={remarks}
                title={t('headings:Remark_And_Signature_Remarks')}
              />
            </Grid> */}
          </Grid>

          <Grid
            item
            xs={12}
            md={4}
            container
            spacing={1}
            direction="column"
            alignItems="center"
            className={classes.signContainer}
          >
            {signs &&
              Array.isArray(signs) &&
              signs.map((item, index) => {
                return (
                  <SignatureContent
                    index={index}
                    signUrl={item.value}
                    title={item.label}
                    showRemoveFileIcon={showRemoveFileIcon}
                    showSaveFileIcon={showSaveFileIcon}
                    isNewlyAdded={item.newlyAdded ? true : false}
                    onCancelClick={removeFile}
                    handleDrop={handleDrop}
                    onSaveClick={onSaveClick}
                  />
                );
              })}
          </Grid>
        </Grid>
        <KenSnackBar
          key={snackbarMessage}
          message={snackbarMessage}
          severity={snackbarSeverity}
          autoHideDuration={4000}
          open={openSnackbar}
          handleSnackbarClose={handleSnackbarClose}
          position="Bottom-Right"
        />
      </Box>
    </>
  );
};
export default RemarkAndSignature;
