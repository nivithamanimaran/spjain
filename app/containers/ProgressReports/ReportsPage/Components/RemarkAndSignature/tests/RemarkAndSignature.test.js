import React from 'react';
import { render, cleanup } from '@testing-library/react';
import ReactDOM from 'react-dom';
import RemarkAndSignature from '../index';
import ThemeProvider from '../../../../../../utils/themeProvider';
import { BrowserRouter as Router } from 'react-router-dom';

afterEach(() => {
  cleanup();
});

describe('Remark and Signature component', () => {
  const div = document.createElement('div');
  it('Remark and Signature component renders correctly', () => {
    ReactDOM.render(
      <ThemeProvider>
        <Router>
          <RemarkAndSignature />
        </Router>
      </ThemeProvider>,
      div
    );

    const { getByTestId } = render(
      <ThemeProvider>
        <Router>
          <RemarkAndSignature />
        </Router>
      </ThemeProvider>
    );

    expect(getByTestId('remark-and-signature-component')).toMatchSnapshot();
  });
});
