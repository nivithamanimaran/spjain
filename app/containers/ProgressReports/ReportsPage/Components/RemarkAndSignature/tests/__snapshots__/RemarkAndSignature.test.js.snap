// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remark and Signature component Remark and Signature component renders correctly 1`] = `
<div
  class="MuiBox-root MuiBox-root-14"
  data-testid="remark-and-signature-component"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-5"
      >
        <div
          class="MuiBox-root MuiBox-root-15 makeStyles-container-5"
          data-testid="remark-content"
        >
          <p
            class="MuiTypography-root makeStyles-headings-3 MuiTypography-body1"
            data-testid="remark-content-title"
          >
            headings:Remark_And_Signature_Details
          </p>
          <div
            class="MuiBox-root MuiBox-root-16 makeStyles-box-10"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-direction-xs-column"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-7"
      >
        <div
          class="MuiBox-root MuiBox-root-17 makeStyles-container-5"
          data-testid="remark-content"
        >
          <p
            class="MuiTypography-root makeStyles-headings-3 MuiTypography-body1"
            data-testid="remark-content-title"
          >
            headings:Remark_And_Signature_Remarks
          </p>
          <div
            class="MuiBox-root MuiBox-root-18 makeStyles-box-10"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-direction-xs-column"
            >
               
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-signContainer-1 MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
    />
  </div>
</div>
`;
